File: 5-minimum_of_three/task.erl
-module(task).

-export([min3/2]).

min3(Count, Values) ->
  min3(Count, Values, []).

min3(_, [], Result) ->
  lists:reverse(Result);
min3(Count, [ValueHead|ValuesTail], Result) ->
  {Value1, Value2, Value3} = ValueHead,
  ValueMin = mini3(Value1, Value2, Value3),
  min3(Count - 1, ValuesTail, [ValueMin|Result]).

mini3(A,B,C) when A < B, A < C ->
  A;
mini3(A,B,C) when B < A, B < C ->
  B;
mini3(A,B,C) when C < A, C < B ->
  C;
mini3(_,_,_) -> 0.

File: 5-minimum_of_three/task.ex
defmodule Task do

  def min3(count, values) do
    min3(count, values, [])
  end

  defp min3(_, [], result) do
    Enum.reverse(result)
  end

  defp min3(count, [{value1, value2, value3}|tail_values], result) do
    value_min = mini3(value1, value2, value3)
    min3(count - 1, tail_values, [value_min|result])
  end
  
  defp mini3(a,b,c) when a < b and a < c, do: a
  defp mini3(a,b,c) when b < a and b < c, do: b
  defp mini3(a,b,c) when c < a and c < b, do: c
  defp mini3(_,_,_), do: 0

end
